opensafely/log_stats.py

```diff
@@ -0,0 +1,156 @@
+#!/usr/bin/python3
+import json
+import re
+from argparse import ArgumentParser
+from datetime import datetime
+from pathlib import Path
+
+DESCRIPTION = "Parse and output cohortextractor-stats logs as JSONL"
+TIMESTAMP_PREFIX_REGEX = r"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}.\d{9}Z"
+COMPLETED_PREFIX_REGEX = r"^Completed successfully"
+KEY_VALUE_REGEX = r"(?<=\s)([^\s=]+)=(.*?)(?=(?:\s[^\s=]+=|$))"
+ACTION_SUMMARY_REGEX = (
+    r"^(state|docker_image_id|job_id|run_by_user|created_at|completed_at):\s(.*)"
+)
+
+
+def add_arguments(parser):
+    parser.add_argument(
+        "--project-dir",
+        help="Project directory (default: current directory)",
+        type=Path,
+        default=".",
+    )
+    parser.add_argument(
+        "--project-name",
+        help="Project name (default: project-dir)",
+        type=str,
+    )
+    parser.add_argument(
+        "--output-file",
+        help="Output json filename (default: log_stats.json)",
+        type=Path,
+        default="log_stats.json",
+    )
+    return parser
+
+
+def _timestamp_for_honeytail(timestamp, ts_format):
+    honeytail_ts_format = "%Y-%m-%dT%H:%M:%S.000000000Z"
+    dt = datetime.strptime(timestamp, ts_format)
+    return dt.strftime(honeytail_ts_format)
+
+
+def parse_log(project_name, current_action, job_id, current_log):
+    current_log_timestamp = re.match(TIMESTAMP_PREFIX_REGEX, current_log).group()
+    current_log = re.sub(r"\s*\n\s*", " ", current_log).strip()
+    log_params = dict(re.findall(KEY_VALUE_REGEX, current_log))
+    ts_format = "%Y-%m-%dT%H:%M:%S"
+    return {
+        "timestamp": current_log_timestamp,
+        "project": project_name,
+        "job_id": job_id,
+        "action": current_action,
+        **log_params,
+    }
+
+
+def parse_stats_logs(raw_logs, project_name, current_action, job_id):
+    for log in raw_logs:
+        yield parse_log(project_name, current_action, job_id, log)
+
+
+def format_summary_stats(project_name, current_action, summary_stats):
+    start_time = summary_stats.get("created_at")
+    end_time = summary_stats.get("completed_at")
+    ts_format = "%Y-%m-%dT%H:%M:%SZ"
+    start_dt = datetime.strptime(start_time, ts_format)
+
+    if start_time and end_time:
+        elapsed = datetime.strptime(end_time, ts_format) - start_dt
+        summary_stats["elapsed_time_secs"] = elapsed.seconds
+        summary_stats["elapsed_time"] = str(elapsed)
+    return {
+        "timestamp": _timestamp_for_honeytail(start_time, ts_format),
+        "project": project_name,
+        "action": current_action,
+        **{k: v for k, v in summary_stats.items()},
+    }
+
+
+def main(project_dir, output_file, project_name=None):
```

Comment:
This comment is going to be unhelpful, but I'm going to make it and reflect on what I can do about it. Anyhow, I found this function _really_ hard to interpret. Partly, that's because I stepped through this PR commit-by-commit, and this function has changed between commits. Partly, that's because I'd not encountered a `cohortextractor-stats` log entry in the wild, so I was missing important context.

Reply:
I can add more comments :)

Reply:
I found it hard to understand intent; I'm still unsure I understand the relationship between `raw_logs` and `current_log`. Am I correct in thinking that `cohortextractor-stats` log entries don't span multiple lines? (But other log entries can.)

Possibly a comment like:

```
Iterate over each line in the log file, collecting cohortextractor-stats log entries in raw_logs.
Finally, parse these log entries into a dict and write to a file.
```

Which isn't quite what's going on, but I think is more-or-less correct 🤔

Reply:
I've tried to add more comments to clarify what this is doing.

Both cohortextractor-stats and other logs can span multiple lines.  We want to keep stats logs but ignore other logs.  `current_log` (which is now renamed to `extracted_log`) is keeping track of the log that we're processing; when we read a new line and it's the start of the next log, we have to store the previous log, and start tracking this one (assuming it's a stats log).  So `current_log` isn't the current line when we read a new line, but could be set to the current line when we process it.

That probably doesn't help at all.

Reply:
The comments are great, thank you 🙂 I hadn't quite appreciated that we're working "one ahead", because of the nature of the log entries.